src/mpid/ch4/src/ch4_recv.h

```diff
@@ -225,20 +230,26 @@ MPL_STATIC_INLINE_PREFIX int MPID_Imrecv(void *buf,
                                          MPIR_Request * message, MPIR_Request ** rreqp)
 {
     int mpi_errno = MPI_SUCCESS;
+
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_IMRECV);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_IMRECV);
 
     if (message == NULL) {
         MPIDI_Request_create_null_rreq(*rreqp, mpi_errno, goto fn_fail);
         goto fn_exit;
     }
+
+    MPIR_Assert(message->kind == MPIR_REQUEST_KIND__MPROBE);
+    message->kind = MPIR_REQUEST_KIND__RECV;
+    *rreqp = message;   /* SHM will override this pointer */
+
 #ifdef MPIDI_CH4_DIRECT_NETMOD
-    mpi_errno = MPIDI_NM_mpi_imrecv(buf, count, datatype, message, rreqp);
+    mpi_errno = MPIDI_NM_mpi_imrecv(buf, count, datatype, message);
 #else
     if (MPIDI_CH4I_REQUEST(message, is_local))
         mpi_errno = MPIDI_SHM_mpi_imrecv(buf, count, datatype, message, rreqp);
     else
-        mpi_errno = MPIDI_NM_mpi_imrecv(buf, count, datatype, message, rreqp);
+        mpi_errno = MPIDI_NM_mpi_imrecv(buf, count, datatype, message);
```

Comment:
shmem mode does create a separate request from `message`. I don't like the asymmetry between netmod and shmem, but I think it's fine for now. We can reevaluate the problem and decide how to bring back symmetry. My hunch is that `rreqp` is unnecessary for shmem as well, in which case we should fix shmem instead of forcing netmod to have an unnecessary argument. We could also keep the netmod API as is and simply not modify the  `rreqp` argument